### PR TITLE
Add functionality to disable plugin via comments

### LINKED ISSUE #17: Add functionality to ignore part of a file (inline comments)
Potentially using something like

`/* postcss-increase-specificity disable-line */`
or
`/* postcss-increase-specificity disable */`
`/* postcss-increase-specificity enable */`

### PR DESCRIPTION
**Pull request by track0x1**
Closes #17 

It only makes sense to support disabling blocks as this plugin only looks at the rule-level (as opposed to line/declaration level).

```css
/* postcss-increase-specificity disable */
.foo {
  content: "can't touch this";
}
/* postcss-increase-specificity enable */
```

Thanks to @MadLittleMods for creating this plugin (it's already been a huge help in a project)!

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -40,6 +42,24 @@ function increaseSpecifityOfRule(rule, opts) {
...
+			return true;
+		} else if (BLOCK_ENABLE.test(prev.text)) {
+			status.disabled = false;
+			return false;
+		}
+	}
+
+	// if no value, use last status.
+	return status.disabled;
+}
```
Comment (MadLittleMods) [inline index.js]: Currently this fails with the following example (let's make sure to add a test),

```css
/* postcss-increase-specificity disable */
/* Some other comment you picked up instead of the one above */
.foo {
	background: #f00;
}
/* postcss-increase-specificity enable */
```

I'm also concerned with other random nodes in between. We should check between the current rule node and the last rule node we walked, `checkForEnableDisableComments(fromNode, toNode)`

File: index.js [hunk 2]
```diff
@@ -40,6 +42,24 @@ function increaseSpecifityOfRule(rule, opts) {
 	}
 }
 
+function isDisabled(node, status) {
```
Comment (MadLittleMods) [inline index.js]: New signature, `checkForEnableDisableComments(fromNode, toNode)`

File: index.js [hunk 3]
```diff
@@ -55,12 +75,14 @@ module.exports = postcss.plugin('postcss-increase-specificity', function(options
 	var opts = objectAssign({}, defaults, options);
 
 	return function(css) {
+		var status = { disabled: false };
```
Comment (MadLittleMods) [inline index.js]: I think I would prefer this sort of setup instead of passing around `status` by reference

```js
var isPluginDisabled = false;

css.walkRules(function(rule) {
	var enableDisableResult = checkForEnableDisableComments(fromNode, toNode);
	isPluginDisabled = (enableDisableResult !== null && enableDisableResult !== undefined) ? enableDisableResult : isPluginDisabled;
        ...
});
```

File: index.js [hunk 4]
```diff
@@ -7,6 +7,8 @@ require('string.prototype.repeat');
 
 var CSS_ESCAPED_TAB = '\\9';
 
+var BLOCK_DISABLE = /postcss-increase-specificity disable/;
+var BLOCK_ENABLE = /postcss-increase-specificity enable/;
```
Comment (MadLittleMods) [inline index.js]: Let's go with these names so we don't confuse blocks/rules and `RE` so we know it is a regex object,

 - `DISABLE_PLUGIN_RE`
 - `ENABLE_PLUGIN_RE`

File: test/test.js [hunk 5]
```diff
@@ -119,4 +119,8 @@ describe('postcss-increase-specificity', function() {
 	it('should not change the descendant rules of @keyframes', function() {
 		return testPlugin('./test/fixtures/keyframes.css', './test/fixtures/keyframes.expected.css');
 	});
+
+	it('should support disabled blocks of rules', function() {
+		return testPlugin('./test/fixtures/disabled-block.css', './test/fixtures/disabled-block.expected.css');
+	});
```
Comment (MadLittleMods) [inline test/test.js:125]: Let's add another test for multiple `disable/enable`

```css
/* postcss-increase-specificity disable */
.noop {
	background: #f00;
}
/* postcss-increase-specificity enable */

.bar {
  color: #ef0;
}

/* postcss-increase-specificity disable */
.noop2 {
	background: #f00;
}
/* postcss-increase-specificity enable */
```

---

Let's add a test for just a `disable` comment on it's own

```css
/* postcss-increase-specificity disable */
.foo {
	background: #f00;
}

.bar {
  color: #ef0;
}

/* some other random comment that should not enable */
.baz {
  color: #040;
}
```

And another one in this same vein

```css
.foo {
	background: #f00;
}

/* postcss-increase-specificity disable */
.bar {
  color: #ef0;
}

.baz {
  color: #040;
}
```

**Review by MadLittleMods — CHANGES_REQUESTED** [review]
Some bugs with multiple random nodes between rules and let's add some more tests 💪 

Thanks for the contribution ❤️

File: index.js (hunk 2 above)
Comment (track0x1) [reply index.js]: I don't think a signature like that makes sense in this case. If we are already holding the previous result in `isPluginDisabled`, then we only have to check above the current node.

I've updated the function name and made it to continue traversing up comments until it finds one that matches OR no more comments are found.

File: test/test.js (hunk 5 above)
Comment (track0x1) [reply test/test.js:125]: Done!

**Comment by track0x1** [comment]
@MadLittleMods Can we get this merged?

File: index.js [hunk 6]
```diff
@@ -40,6 +42,25 @@ function increaseSpecifityOfRule(rule, opts) {
...
+		if (prev.type !== 'comment') return null;
+		if (DISABLE_PLUGIN_RE.test(prev.text)) {
+			return true;
+		} else if (ENABLE_PLUGIN_RE.test(prev.text)) {
+			return false;
+		}
+	}
+
+	return null;
+}
```
Comment (MadLittleMods) [inline index.js:63]: This function currently checks from the given node upwards through any comments. But there can be other types of nodes in the way. Let's be sure to add a test

```css
/* postcss-increase-specificity disable */

random-decl-here: 'why';

.foo {
	background: #f00;
}
/* postcss-increase-specificity enable */

.bar {
	color: #ef0;
}
```

I still think we should `checkForEnableDisableComments(fromNode, toNode)` and `toNode.prev(...)` loop to `toNode`, stopping at the first matching disable/enable comment.

File: test/test.js [hunk 7]
```diff
@@ -119,4 +119,24 @@ describe('postcss-increase-specificity', function() {
...
+		return testPlugin('./test/fixtures/multiple-disabled-blocks.css', './test/fixtures/multiple-disabled-blocks.expected.css');
+	});
+
+	it('should support an entirely disabled stylesheet', function() {
+		return testPlugin('./test/fixtures/disabled-stylesheet.css', './test/fixtures/disabled-stylesheet.expected.css');
+	});
+
+	it('should support a partially disabled stylesheet', function() {
+		return testPlugin('./test/fixtures/partially-disabled-stylesheet.css', './test/fixtures/partially-disabled-stylesheet.expected.css');
+	});
```
Comment (MadLittleMods) [inline test/test.js:141]: Let's add some tests for comments inside nested rules

```css
.foo {
	background: #f00;

	/* postcss-increase-specificity disable */
	.nested-foo {
		color: #00a;
	}
	/* postcss-increase-specificity enable */
}

.bar {
	color: #ef0;
}
```

Also do we expect this to disable the outer rules 🤷‍♀️? Let's add a test for this but I'm up for this case to disable everything below regardless of where it's nested.

```css
.foo {
	background: #f00;

	/* postcss-increase-specificity disable */
	.nested-foo {
		color: #00a;
	}
}

.bar {
	color: #ef0;
}
```

**Review by MadLittleMods — CHANGES_REQUESTED** [review]
I found some new edge cases we should handle

File: test/test.js (hunk 7 above)
Comment (track0x1) [reply test/test.js:141]: @MadLittleMods I thought at the `postcss` level we are dealing with CSS and not Sass?

File: index.js (hunk 6 above)
Comment (track0x1) [reply index.js:63]: @MadLittleMods There is no genuine use case for a declaration outside of a class; that's not valid CSS.

Comment (MadLittleMods) [reply index.js:63]: @track0x1 Just because it isn't normal CSS, doesn't mean that it come up as it parses fine with PostCSS. We should handle this situation.

File: test/test.js (hunk 7 above)
Comment (MadLittleMods) [reply test/test.js:141]: @track0x1 We are dealing with whatever the PostCSS parser can handle

File: index.js (hunk 6 above)
Comment (track0x1) [reply index.js:63]: Ok, so I think we can safely ignore declarations.

File: test/test.js (hunk 7 above)
Comment (track0x1) [reply test/test.js:141]: I think this request is out of scope for this PR as currently the plugin doesn't handle nested rules.

File: index.js [hunk 8]
```diff
@@ -40,6 +42,25 @@ function increaseSpecifityOfRule(rule, opts) {
...
 
+/**
+ * Checks for the presence of enable/disable comments up to the previous rule.
+ * @param  {node} node 		PostCSS Node
+ * @return {Null|Boolean} Null if no matching comment; otherwise boolean (toggle status)
+ */
+function checkForEnableDisableComments(node) {
+	if (!node) return null;
+
+	for (var prev = node; prev = prev.prev(); ) {
```
Comment (MadLittleMods) [inline index.js:53]: This for-loop variation with the `final-expression` as the `condition` threw me off a bit. The statement even runs for the first item which can be unexpected if you were reasoning about it as an iterator (`final-expression`).

Let's update this to the more familiar 3 part syntax,

```
for (var prev = node.prev(); prev !== prev.prev(); prev = prev.prev()) {
```

File: index.js (hunk 6 above)
Comment (MadLittleMods) [reply index.js:63]: This logic is still flawed. Consider the following example (let's add a test as well),

```css
/* postcss-increase-specificity disable */

.foo {
	background: #f00;
}
/* postcss-increase-specificity enable */

@some-at-rule {

}

.bar {
	color: #ef0;
}
```

---

Please update this method to be generic to handle any type of node in between, even future unknown types.

I think we should go with this,

```js
/**
 * Checks for the presence of enable/disable comments
 * starting from the `fromNode` upwards to the previous rule (`toNode`).
 * @param  {fromNode} node 		PostCSS Node
 * @param  {toNode} node 		PostCSS Node
 * @return {Null|Boolean} Null if no matching comment; otherwise boolean (toggle status)
 */
function checkUpwardsForEnableDisableComments(fromNode, toNode) {
	if (!fromNode) return null;

	for (var prev = fromNode.prev(); (prev !== toNode && prev !== undefined); prev = prev.prev()) {
		if (DISABLE_PLUGIN_RE.test(prev.text)) {
			return true;
		} else if (ENABLE_PLUGIN_RE.test(prev.text)) {
			return false;
		}
	}

	return null;
}
```

File: test/test.js [hunk 9]
```diff
@@ -119,4 +119,28 @@ describe('postcss-increase-specificity', function() {
...
+
+	it('should support disabled blocks of rules with declarations in between', function() {
+		return testPlugin('./test/fixtures/disabled-block-extra-declaration.css', './test/fixtures/disabled-block-extra-declaration.expected.css');
```
Comment (MadLittleMods) [inline test/test.js:144]: wdyt of this name to describe the situation better? `extraneous-declaration-outside-rule-between-disable-comment`

File: test/test.js (hunk 7 above)
Comment (MadLittleMods) [reply test/test.js:141]: Let's add a comment in the readme about this known limitation/design-decision.

> Currently, these disable/enable comments can not be nested inside a rule. Only comments in the global scope will be taken into account.

**Review by MadLittleMods — CHANGES_REQUESTED** [review]
Let's make `checkUpwardsForEnableDisableComments(fromNode, toNode)` generic to handle any type of node

**Comment by MadLittleMods** [comment]
Closing as this has gone stale and review not addressed.
